src/module/chat.ts

```diff
@@ -1,50 +1,91 @@
 import { SR5Actor } from './actor/SR5Actor';
 import { SR5Item } from './item/SR5Item';
 import Template from './template';
-import ModList = Shadowrun.ModList;
-import BaseValuePair = Shadowrun.BaseValuePair;
 import DamageData = Shadowrun.DamageData;
 import AttackData = Shadowrun.AttackData;
-import LabelField = Shadowrun.LabelField;
-import { FLAGS, SYSTEM_NAME } from './constants';
-import { PartsList } from './parts/PartsList';
+import {CORE_FLAGS, CORE_NAME, FLAGS, SYSTEM_NAME} from './constants';
+import {ShadowrunRoll, ShadowrunRoller, Test} from "./rolls/ShadowrunRoller";
+import DrainData = Shadowrun.DrainData;
 
-export type TemplateData = {
+// Simple card text messages
+interface ItemChatMessageOptions {
+    actor: SR5Actor
+    item: SR5Item
     header: {
-        name: string;
-        img: string;
-    };
-    tokenId?: string;
-    dice?: Die[];
-    parts?: ModList<number>;
-    limit?: BaseValuePair<number> & LabelField;
-    testName?: string;
-    actor?: SR5Actor;
-    item?: SR5Item;
-    attack?: AttackData;
-    incomingAttack?: AttackData;
-    incomingDrain?: LabelField & {
-        value: number;
-    };
-    hits?: number;
-    soak?: DamageData;
-    tests?: {
-        label: string;
-        type: string;
-    }[];
+        name: string
+        img: string
+    }
+    description: object
+    tests?: Test[]
+}
+
+interface RollChatMessageOptions {
+    actor?: SR5Actor
+    target?: Token
+
+    item?: SR5Item
+
+    name?: string
+    img?: string
+    title: string
+    description?: object
+
+    rollMode?: keyof typeof CONFIG.dice.rollModes
+    previewTemplate?: boolean
+
+    attack?: AttackData
+    incomingAttack?: AttackData
+    incomingDrain?: DrainData
+    incomingSoak?: DamageData
+    tests?: Test[];
+}
+
+interface ItemChatTemplateData {
+    actor: SR5Actor
+    item: SR5Item
+    header: {
+        name: string
+        img: string
+    }
+    description: object
+    tests?: Test[]
+}
+
+interface RollChatTemplateData {
+    actor?: SR5Actor
+    header: {
+        name: string
+        img: string
+    }
+    tokenId?: string
+    target?: Token
+    item?: SR5Item
+
+    title: string
     description?: object;
-    previewTemplate?: boolean;
-    rollMode?: keyof typeof CONFIG.Dice.rollModes;
-};
 
-export const createChatData = async (templateData: TemplateData, roll?: Roll) => {
+    roll: ShadowrunRoll
+
+    rollMode: keyof typeof CONFIG.dice.rollModes
+    previewTemplate?: boolean
+
+    attack?: AttackData
+    // TODO: group 'incoming' with type field instead of multiple incoming types.
```

Comment:
I fully agree with this. A Discriminated Union in Typescript would work here nicely